CLAM/plugins/continuousExcitationSynthesizer/ContinuousExcitationSynthesizer.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH CONTINUOUSEXCITATIONSYNTHESIZER "1" "March 2009" "ContinuousExcitationSynthesizer version 1.3.1~svn12688" "User Commands"
.SH NAME
ContinuousExcitationSynthesizer \- OSC controlled synthesizer based on Spectral Modelling techniques
.SH SYNOPSIS
.B ContinuousExcitationSynthesizer
\fIinputSDIFDirectory \fR[\fI(osc | inputScore) \fR[ \fIoutput.wav \fR[ \fIsynthesis.xml \fR] ] ]
.SH DESCRIPTION
If osc is given, the application will expect an osc stream
with this message format: /ebowSynthesizer pitch amplitude
to be sent to port 7000. Pitch should be in Hertz and amplitude
should be from 0\-1.
.PP
If an input score is given, the application will expect the
score to have the same format as the files in testData, i.e.
pitch amplitude
pitch amplitude ...
.PP
One synthesis frame will be calculated for each OSC message.
or each line from a score file. So, given a hopsize of 256,
you'll need to generate about 172 control events per second.
.SH AUTHOR
Written by Greg Kellum
.SH COPYRIGHT
Copyright \(co 2009, Universitat Pompeu Fabra.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Licence GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>


CLAM/CHANGES

2008-??-?? CLAM 1.3.1 SVN $Revision$
''
 * Typed controls: Controls with other type than floats.
 * Regular in controls have callback (deprecates InControlTmpl)
 * Networks can pause without stopping the backend.
 * Processing::Get{In,Out}{Control,Port}() overloaded for indexes and names.
 * Processing::GetN{In,Out}{Controls,Ports}() to get the number of connectors.
 * Processing::Get{In,Out}{Controls,Ports}() are now protected methods.

2008-08-11 CLAM 1.3.0
'Shooting flying plugins release'
 * SVN Revision: 11960.
 * Integrates first wave of GSoC 2008 contributions from the students
   Natanael Olaiz, Wang Jun, Pawel Bartkiewicz, Francisco Tufro,
   and Yushen Han.
 * Native windows compilation (using mingw) working without patches (GSoC:jun)
 * A simpler and more effective API to export LADSPA plugins from networks
   and isolated processings.
 * Template based processing/plugin code generator scripts and frontend.
 * Many changes to the Network allowing to do cut and paste, embeded
   positions...
 * Spatialization plugin:
   * ComplexSpectrumInverter: Inverts the phase of a ComplexSpectrum
   * Click-less impulse response switching,
   * Choreography sequencer to calculate simulated impulses on the fly
     instead of using a dense grid of precomputed IRs.
     (requires a non-public binary but you can plug-in a different one)
   * Spherical2CartesianControl: Converts radix,azimut,elevation controls
     to x,y,z
 * Sample by sample processings now are a plugin.
 * New processings:
   * AudioFileMemoryLoader: an audio file reader without locks and with seek
     but loading the full audio into memory (a ringbuffer based on the way).
   * Resampler: Resamples audio streams using libsamplerate (not stable)
 * TickExtractor example working again (thanks 
 * Fix: NetEditor Windows installer should not install .def and .lib for
   plugins
 * Voice2MIDI crosscompiled windows binaries (no installer yet)
 * Old AudioFileInput/Output definitely removed, remaining old code adapted.
 * FormantTracker and TickExtractor examples working with current version

2008-02-05 CLAM 1.2.0
'GSoCket Plugged-in release'
 * SVN Revision: 11171
 * Major changes summary:
   * Integrates GSoC students work. Thanks to Hernan Ordiales
      Bennet Kolasinsky, Greg Kellum, Andreas Calvo, Roman Goj,
      Abe Kazemzadeh and Google Inc.
   * It also integrates work done last month within the
     Acoustics team on the Barcelona Media center.
   * CLAM plugins: You can load your own processings and
     data types in an existing CLAM application such the
     Network Editor. (GSoC:acalvo)
   * Windows binaries crosscompile from Linux
 * NetworkEditor enhancements:
   * Enhanced online tutorial
   * Token data types expose a color to show in NetworkEditor
   * Port tooltips show also the data type
   * Suitable monitors are available at port context menu
   * Suitable controlers are available at control context menu
     or double clicking
   * The processing is automatically populated with available
     processings. (GSoC:acalvo)
   * Dialog to add processings has completion
   * New spectrogram view (GSoC:bkolasinsky)
 * Realtime port to some spectral processings and prototypes:
    * Harmonizer, pitch-discretization and morph (GSoC:hordia)
 * Chord analysis modularity, configurability and postprocessing
   improved (GSoC:rgoj)
 * Anotator enhancements (GSoC:bkolasinsky):
   * Multiple segmentation and low-level features panes
   * BarGraph is available as Instant View
   * Rulers are positioned at round numbers
   * Rulers have subticks for smaller not numbered divisions
   * Schema browsers shows Scopes (ie. Frame) as subtrees of
     their defining attribute (ie. Song::FrameDivision)
 * First plugins available:
   * guitardistortion: Cool guitar effects with prototyper
     interface. (GSoC:hordia)
   * continuousExcitationSynth: An SMS based synthesizer.
     (GSoC:gkellum)
   * speech: Experimental processings for speech synthesis
     and analysis (GSoC:akazemzadeh)
   * spacialization: 3D audio processing including
     * New spectral processings with simplifyied/explicit
       polar-complex representation, including windowing,
       fft, ifft, overlap-and-add and representation
       converters.
     * Real-time low latency convolution
     * Grid and polar impulse response database fetcher
     * Exibition processings to convert p+v format to
       5:1, N-Surround and other exibition formats.
     * Several room databases are available, check the wiki.
 * Build system:
   * Windows binaries obtained by crosscompiling from linux
   * Windows external libs are deployed now as a posix tree
   * Relying on pkg-config to link dependencies
   * VisualC bits droped
   * Installing plugins
   * OSX uses macported Qt instead of Trolltech binaries

2007-06-07 CLAM 1.1.0
'More eye-candy, please'
 * SVN Revision: 10216
 * New concise task oriented documentation (on the wiki)
 * Loading of faust (ladspa) modules
 * FLTK fully dropped
 * Qt3 based vmqt dropped (some files moved to SMSTools)
 * '_support' ending for scons configure options removed:
   with_jack_support, with_ladspa_support, with_osc_support
   became with_jack, with_ladspa, with_osc
 * Examples clean up
 * Realtime port to some processings:
   * SMSGenderChange, SMSPitchShift, SMSShapeShift
 
2007-03-20 CLAM 1.0.0
'The Berlin release'
 * SVN Revision: 9881
 * Funtional Ladspa host

2007-03-18 CLAM 0.99.0
'That will be CLAM 1.0 release'
 * SVN Revision: 9864
 * New binary in NetworkEditor/ called OfflinePlayer to batch-process
   clam network files given an input and output audio file.
 * Fixes in the modconf to work in suse (thanks Toni)
 * Updated Fedora Core packaging
 * PortAudio network player backend:
   * PortAudio available in all platforms, including Linux.
   * Network player gets priority boost in Linux.
   * Discarded OSS API so in linux it uses Alsa (#Closes 200).
   * Xruns are notified on the console.

trunk/CLAM/plugins/continuousExcitationSynthesizer/ContinuousExcitationSynthesizer.cxx

#include <CLAM/XMLStorage.hxx>
#include <CLAM/SMSSynthesis.hxx>
#include <CLAM/SMSSynthesisConfig.hxx>
#include <CLAM/SinusoidalSynthesis.hxx>
#include <CLAM/Audio.hxx>
#include "Processing.hxx"
#include "SDIFInConfig.hxx"
#include "SDIFFileReader.hxx"
#include "BufferedSDIFFileReader.hxx"
#include "SDIFDatabase.hxx"
#include "MonoAudioFileWriter.hxx"
#include "MonoAudioFileWriterConfig.hxx"
#include "Spectrum.hxx"
#include "AudioManager.hxx"
#include "AudioIO.hxx"
#include "AudioOut.hxx"
#include "AudioSink.hxx"
#include "TopLevelProcessing.hxx"
#include "SMSFreqShift.hxx"
#include "TestScore.hxx"
#include "ControlStreamSegmentator.hxx"
#include <CLAM/CLAMVersion.hxx>

#ifdef WITH_OSC
#include "ContinuousExcitationOSCSource.hxx"
#endif

/**
*  This example shows off an implementation of a synthesizer that uses a database
*  of SDIF files to perform real-time Spectral Modeling Synthesis.
*
*  To generate an SDIF file to use with this example, see the example WavToSDIF.
*
*  To run the example, type ./SDIFToWavStreamingExe from the examples directory.
*  The application will print out a help message telling you which files it needs in which order.
*
*  \see {http://www.cnmat.berkeley.edu/SDIF/}
*  \see {http://www.mtg.upf.edu/sms/}
*
*  \author Greg Kellum [gkellum@iua.upf.edu] 7/15/2007
*  \since  CLAM v1.1
*/
namespace CLAM
{

	void SynthesizeToSpeakers(SDIFDatabase& sdifDatabase, SinusoidalSynthesis& synthesis, 
								ContinuousExcitationControlSource* aControlScore, MonoAudioFileWriter* audioWriter)
	{
        if (audioWriter == NULL)
            std::cout << "Playing audio to speakers." << std::endl;
        else
            std::cout << "Playing audio to speakers and writing audio to file." << std::endl;

		CLAM_ACTIVATE_FAST_ROUNDING;

		unsigned int buffersize = 1024;
		int samplerate = 44100;

        AudioManager audioManager(samplerate,2048);
		AudioIOConfig outLCfg;
		outLCfg.SetDevice("rtaudio:default");
		outLCfg.SetChannelID(0);
		outLCfg.SetSampleRate(samplerate);

		AudioIOConfig outRCfg;
		outRCfg.SetDevice("rtaudio:default");
		outRCfg.SetChannelID(1);
		outRCfg.SetSampleRate(samplerate);

		CLAM::SMSFreqShift smsFreqShift;

		CLAM::AudioOut outL(outLCfg);
		CLAM::AudioOut outR(outRCfg);
		//CLAM::AudioSink out;

        audioManager.Start();

		smsFreqShift.Start();
		synthesis.Start();
		if (audioWriter != NULL)
		{
			audioWriter->Start();
		}

        outL.Start();
        outR.Start();
        
		ControlStreamSegmentator theSegmentator;

		int counter = 0;
		VoiceId theVoiceId;
		float thePitch;
		float theAmplitude;
		// make this true to have the given score file loop
		aControlScore->SetFeedForever(false);
		while ( aControlScore->GetControlValues(thePitch, theAmplitude) )
		{
			//std::cout << "pitch: " << thePitch << ", amplitude: " << theAmplitude << std::endl;
			
			// TODO do you really need VoiceIds? Wouldn't it be enough to simply
			// tell the sdifDatabase to read from the attack? 
			bool isNewVoice = theSegmentator.IsNewVoice(thePitch, theAmplitude);
			if ( isNewVoice )
			{
				sdifDatabase.DeleteVoice(theVoiceId);
				theVoiceId = sdifDatabase.CreateVoice();
//				std::cout << "Oraison: creating new voice" << std::endl;
			}
			
			Frame* aFrame = sdifDatabase.SelectFrameForVoice(theVoiceId, thePitch, theAmplitude);

			//std::cout << "counter: " << counter << ", pitch: " << thePitch << ", amplitude: " << theAmplitude << std::endl;
			bool isSynthesizeSinusoidsAndResidual = false;
			if ( !synthesis.Do( *aFrame ) )
                continue;

			if (audioWriter != NULL)
			{
				audioWriter->Do( aFrame->GetSynthAudioFrame() );
			}

            outL.Do( aFrame->GetSynthAudioFrame() );
            outR.Do( aFrame->GetSynthAudioFrame() );
            
			counter++;
		}

		sdifDatabase.DeleteVoice(theVoiceId);

		if (audioWriter != NULL)
		{
			audioWriter->Stop();
		}

        outL.Stop();
        outR.Stop();

		synthesis.Stop();
		smsFreqShift.Stop();

		CLAM_DEACTIVATE_FAST_ROUNDING;

		std::cout << "Computed audio for " << counter << " frames." << std::endl;
	}

} //namespace CLAM

int usage()
{
	std::cout <<
		"Usage: ContinuousExcitationSynthesizer inputSDIFDirectory [(osc | inputScore) [ output.wav [ synthesis.xml ] ] ]\n"
		"\n"
		"This example shows off an implementation of a synthesizer\n"
		"that uses a database of SDIF files to perform real-time\n"
		"Spectral Modeling Synthesis.\n"
		"\n"
		"If 'osc' option is given, the application will expect an osc stream\n"
		"with this message format: /ebowSynthesizer pitch amplitude\n"
		"to be sent to port 7000. Pitch should be in Hertz and amplitude\n"
		"should be from 0-1.\n"
		"\n"
		"If an input score is provided, the application will expect the\n"
		"score to have the same format as the files in testData, i.e.\n"
		"pitch amplitude\n"
		"pitch amplitude ... \n"
		"\n"
		"One synthesis frame will be calculated for each OSC message.\n"
		"or each line from a score file. So, given a hopsize of 256,\n"
		"you'll need to generate about 172 control events per second.\n"
		"\n"
		"You can obtain a sample database at:\n"
		"http://clam-project.org/clam_data/trunk/OboeSDIF/\n"
		"Alternatively, databases can be generated by using WavToSDIF and\n"
		"SDIFToWavStreaming programs available as CLAM framework\n"
		"examples. The first one will convert a regular wave into\n"
		"an SDIF file and the second will prepare a set of SDIF files\n"
		"to build an SMS synthesizer database.\n"
		"\n"
		<< std::endl;
		return 1;
}
int version()
{
	std::cout << 
		"ContinuousExcitationSynthesizer version " << CLAM::GetFullVersion() << "\n"
        "Copyright (C) 2009, Universitat Pompeu Fabra.\n"
        "This is free software: you are free to change and redistribute it.\n"
        "There is NO WARRANTY, to the extent permitted by law.\n"
        "Licence GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>\n"
        "\n"
		"Written by Greg Kellum\n"
		<< std::endl;
	return 1;
}
int main(int argc,char** argv)
{
	bool isPlayOnSpeakers = false;
	const char *sdifDirectory, *scoreFile, *wavoutput, *xmlconfig;
	for (int i=0; i<argc; i++)
	{
		if (argv[i]==std::string("--version")) return version();
		if (argv[i]==std::string("--help")) return usage();
		if (argv[i]==std::string("-h")) return usage();
	}
	switch (argc)
	{
	case 2:
	case 3:
	case 4:
	case 5:
		sdifDirectory=argv[1];
		scoreFile=(argc < 3 )?"osc":argv[2];
		wavoutput=(argc < 4 )?"speakers":argv[3];
		xmlconfig=(argc < 5 )?"synthesis.xml":argv[4];
		isPlayOnSpeakers = (strcmp(wavoutput, "speakers") == 0) ? 1 : 0;
		break;
	default:
		return usage();
	}

	/** Analysis configuration */
	CLAM::SMSSynthesisConfig SynthesisConfig;

	/** Internal Processings used for analysis and synthesis */
	CLAM::SinusoidalSynthesis Synthesis;

	CLAM::ContinuousExcitationControlSource* aControlScorePtr;
	if ( strcmp(scoreFile, "osc") != 0 )
	{
		CLAM::TestScore* aTestScorePtr = new CLAM::TestScore();
		bool result = aTestScorePtr->LoadScore(scoreFile);
		aControlScorePtr = aTestScorePtr;
	}
	else
	{
#ifdef WITH_OSC
		aControlScorePtr = new CLAM::ContinuousExcitationOSCSource(std::string("/ebowSynthesizer"));
#else
		std::cout << "OSC is not enabled. Loading dummy score rather than activating OSC." << std::endl;
		aControlScorePtr = new CLAM::TestScore();
#endif
	}

	/****************************************\
	 * Configure Synthesis				  *
	\****************************************/
	// SMS Analysis configuration
	CLAM::XMLStorage::Restore(SynthesisConfig, xmlconfig);
	// if you'd like to use your own configuration, either edit the XML file
	// or create a SynthesisConfig object by hand and dump its contents
	// to file for later use
	//XMLStorage::Dump(AnalConfig,"SMSAnalysisConfig",xmlconfig);
	Synthesis.Configure(SynthesisConfig);

	/****************************************\
	 * Load SDIF							*
	\****************************************/
//		CLAM::SDIFInConfig cfg;
//		cfg.SetEnableFundFreq(false);
//		cfg.SetEnablePeakArray(true);
//		cfg.SetEnablePeakArray(true);
	std::string directoryName = std::string(sdifDirectory);
	CLAM::SDIFDatabase theSDIFLoader;
	bool isResidualEnabled = false;
	theSDIFLoader.LoadProgram( directoryName, isResidualEnabled );

	//CLAM::Thread threadPtr;
	//SDIFLoader.LoadFramesIntoBufferOnThread( &threadPtr );

	/****************************************\
	 * Configure AudioFileWriter			*
	\****************************************/
	CLAM::MonoAudioFileWriter audioWriter;
	CLAM::MonoAudioFileWriterConfig writercfg;

	const std::string outputFile(wavoutput);
	writercfg.SetTargetFile(outputFile);
	audioWriter.Configure( writercfg );

	/****************************************\
	 * Let's now synthesize the audio	   *
	\****************************************/
	if ( strcmp(wavoutput, "speakers") == 0 )
		SynthesizeToSpeakers(theSDIFLoader, Synthesis, aControlScorePtr, NULL);
	else
		SynthesizeToSpeakers(theSDIFLoader, Synthesis, aControlScorePtr, &audioWriter);

	delete aControlScorePtr;

	std::clog << "Finished successfully!\n";
	return 0;
}

